tavern/mqtt.py

```diff
@@ -0,0 +1,228 @@
+        self._client = paho.Client(**self._client_args)
+        self._client.enable_logger()
+
+        if self._auth_args:
+            self._client.username_pw_set(**self._auth_args)
+
+        self._client.on_message = self._on_message
+
+        if self._enable_tls:
+            self._client.tls_set(**self._tls_args)
+
+        # Arbitrary number, could just be 1 and only accept 1 message per stages
+        # but we might want to raise an error if more than 1 message is received
+        # during a test stage.
+        self._message_queue = Queue(maxsize=10)
+        self._userdata = {
+            "queue": self._message_queue,
+        }
+        self._client.user_data_set(self._userdata)
+
+    @staticmethod
+    def _on_message(client, userdata, message):
+        """Add any messages received to the queue
+
+        Todo:
+            If the queue is faull trigger an error in main thread somehow
+        """
+        # pylint: disable=unused-argument
+
+        logger.info("Received mqtt message on %s", message.topic)
+
+        try:
+            userdata["queue"].put(message)
+        except Full:
+            logger.exception("message queue full")
+
+    def message_received(self, topic, payload, timeout=1):
+        """Check that a message is in the message queue
+
+        Args:
+            topic (str): topic message should have been on
+            payload (str, dict): expected payload - can be a str or a dict...?
+            timeout (int): How long to wait before signalling that the message
+                was not received.
+
+        Returns:
+            bool: whether the message was received within the timeout
+
+        Todo:
+            Allow regexes for topic names? Better validation for mqtt payloads
+        """
+
+        time_spent = 0
+
+        while time_spent < timeout:
+            t1 = time.time()
+
+            try:
+                msg = self._message_queue.get(block=True, timeout=timeout)
+            except Empty:
+                time_spent += timeout
+            else:
+                time_spent += time.time() - t1
+                if msg.payload != payload and msg.topic != topic:
+                    # TODO
+                    # Error?
+                    logger.warning("Got unexpected message in '%s' with payload '%s'",
+                        msg.topic, msg.payload)
+                else:
+                    logger.warning("Got expected message in '%s' with payload '%s'",
+                        msg.topic, msg.payload)
+
+                    return True
+
+        logger.error("Message not received after %d seconds", timeout)
+        return False
+
+    def publish(self, topic, payload, *args, **kwargs):
+        """publish message using paho library
+        """
+        logger.debug("Publishing on '%s'", topic)
+        self._client.subscribe(topic, qos=1)
+
+        if isinstance(payload, (dict, list)):
+            payload = json.dumps(payload)
```

Comment:
Can't assume JSON, we should at least have a key in the yaml for this, even if only JSON is supported without an external plugin.

Reply:
we could just have a `json` key mutually exclusive with a `payload` key where json is automatically dumped

Reply:
It now has a 'json' key alongside 'payload' which is automatically dumped/loaded as json and used as the payload.